src/engine/ExecuteUpdate.cpp

```diff
@@ -46,61 +46,101 @@ std::pair<std::vector<ExecuteUpdate::TransformedTriple>, LocalVocab>
+    if (tc.isVariable()) {
+      AD_CORRECTNESS_CHECK(variableColumns.contains(tc.getVariable()));
+      return variableColumns.at(tc.getVariable()).columnIndex_;
     }
+    if (!(tc.isLiteral() || tc.isIri())) {
+      return tc.toValueIdIfNotString(&encodedIriManager).value();
+    }
+    auto found = lookupMap.find(tc);
+    AD_CORRECTNESS_CHECK(found != lookupMap.end());
+    return found->second;
   };
-  Id defaultGraphIri = [&transformSparqlTripleComponent] {
-    IdOrVariableIndex defaultGraph = transformSparqlTripleComponent(
+  Id defaultGraphIri = [&lookupTc] {
+    const IdOrVariableIndex defaultGraph = lookupTc(
         ad_utility::triple_component::Iri::fromIriref(DEFAULT_GRAPH_IRI));
     AD_CORRECTNESS_CHECK(std::holds_alternative<Id>(defaultGraph));
     return std::get<Id>(defaultGraph);
   }();
-  auto transformGraph = [&vocab, &localVocab, &defaultGraphIri,
-                         &variableColumns, &encodedIriManager](
-                            SparqlTripleSimpleWithGraph::Graph graph) {
+  auto lookupGraph = [&defaultGraphIri, &variableColumns, &lookupTc](
+                         const SparqlTripleSimpleWithGraph::Graph& graph) {
     return std::visit(
         ad_utility::OverloadCallOperator{
             [&defaultGraphIri](const std::monostate&) -> IdOrVariableIndex {
               return defaultGraphIri;
             },
-            [&vocab, &localVocab,
-             &encodedIriManager](const ad_utility::triple_component::Iri& iri)
-                -> IdOrVariableIndex {
-              return TripleComponent(iri).toValueId(vocab, localVocab,
-                                                    encodedIriManager);
+            [&lookupTc](const ad_utility::triple_component::Iri& iri) {
+              return lookupTc(TripleComponent(iri));
             },
             [&variableColumns](const Variable& var) -> IdOrVariableIndex {
               AD_CORRECTNESS_CHECK(variableColumns.contains(var));
               return variableColumns.at(var).columnIndex_;
             }},
         graph);
   };
-  auto transformSparqlTripleSimple =
-      [&transformSparqlTripleComponent,
-       &transformGraph](SparqlTripleSimpleWithGraph triple) {
-        return std::array{transformSparqlTripleComponent(std::move(triple.s_)),
-                          transformSparqlTripleComponent(std::move(triple.p_)),
-                          transformSparqlTripleComponent(std::move(triple.o_)),
-                          transformGraph(std::move(triple.g_))};
+  auto transformTriple =
+      [&lookupTc, &lookupGraph](const SparqlTripleSimpleWithGraph& triple) {
+        return std::array{lookupTc(triple.s_), lookupTc(triple.p_),
+                          lookupTc(triple.o_), lookupGraph(triple.g_)};
       };
-  return {
-      ad_utility::transform(std::move(triples), transformSparqlTripleSimple),
-      std::move(localVocab)};
+  return {ad_utility::transform(triples, transformTriple),
```

Comment:
Why is the ref made const? (You could've just turned it into a normal type without a reference if the && is the problem). This adds many additional copies that were previously avoided.